File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void combination(vector<int>& candidates,int target,vector<vector<int>>&ans,vector<int>&ds,int index){
        if(target == 0)
        {
            ans.push_back(ds);
            return;
        }
        
        if(index >= candidates.size()){
            return;
        }
        
        //pick element
        if(candidates[index] <= target){
        ds.push_back(candidates[index]);
        combination(candidates,target - candidates[index],ans,ds,index);
        
        ds.pop_back();
        }
        //not pick
        combination(candidates,target,ans,ds,index+1);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>ans;
        vector<int>ds;
        
        combination(candidates,target,ans,ds,0);
        
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private void combination( List<List<Integer>>ans,List<Integer>v, int[] candidates,int target,int ind){
        if(ind == candidates.length)
        {
            if(target == 0){
                ans.add(new ArrayList<>(v));
            }
            return;
        }
        if(candidates[ind]<=target){
            v.add(candidates[ind]);
            combination(ans,v,candidates,target-candidates[ind],ind);
            v.remove(v.size()-1);
            
     }
        combination(ans,v,candidates,target,ind+1);
        
    }
        
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        combination(ans,new ArrayList<>(),candidates,target,0);
        return ans;
    }
}